### DISCUSSION TITLE
Laravel doesn't support subfolder website (https://mydomain.com/website) ?

### DISCUSSION THREAD

**Discussion by jrbecart** [laravel/framework#34388] [General]
Hi @taylorotwell ,

Sorry to ping you here, but I am (along with others) very confused right now.

It looks like Laravel doesn't support website in a subfolder like https://mydomain.com/website
This is very surprising as this is used in many organizations and institutions almost everywhere.

Many issues were closed saying that Laravel doesn't support this, referring to the doc here: https://laravel.com/docs/7.x/installation#directory-configuration
But there is no explanation here, root of the web directory on the web server and subfolder website are related but two different things.
There is no issue having Lavarel served out of the root of the "web directory" with: a domain, a subdomain or a subfolder website, because it's a web server configuration (NGINX, Apache, etc.).

So the recommendation is fine but doesn't apply to the type of website (domain/subdomain/subfolder)

If it's true that you don't support subfolders websites (https://mydomain.com/website), can you give the reason?
And if so then it should be clearer and added in the documentation.

Thanks for your feedback on this!

Regards,

**Comment by driesvints** [comment]
Hey @jrbecart. The reason is right there in the docs: 

> Attempting to do so could expose sensitive files present within your application.

**Reply by jrbecart** [reply]
Hi @driesvints 

Please read carefully, I think you didn't get it.

You can have your server setup with this true:
Laravel should always be served out of the root of the "web directory" configured for your web server. You should not attempt to serve a Laravel application out of a subdirectory of the "web directory".
**AND** your website in https://mydomain.com/website
**because** it depend on how is configured your web server

> **root of the web directory on the web server and subfolder website are related but two different things**

**Comment by kangzoel** [comment]
Because it's a Laravel app I think 🤔 
In a Laravel app, whatever path you type after the app root will be handled by the router.

If you want to serve another code/ framework, just point a subdomain into that.

**Reply by jrbecart** [reply]
No it perfectly works, it's not the point here. 

Some says it's not supported or recommended, referencing to the doc where it's not clear and it's not explained why. The doc should clearly mention that subfolder websites are not supported with a good explanation. 

I am sure anyone with web server configuration knowledge understand what is a subfolder website and will not understant where is the problem with Laravel specificly, as any other framework/CMS support it (Drupal, Wordpress, OctoberCMS [laravel], etc.)

Is there any person with Nginx, Apache or other web server experience to explain how a subdirectory website works and why Laravel not supporting it? (Other than reffering to the doc)

**Reply by kangzoel** [reply]
What do you mean by subdirectory website? I don't get it.

**Reply by kangzoel** [reply]
If you place the laravel folder under /path/to/web/directory/laravel  (which is not recommended), it can be accessed in http://example.com/laravel/public

**Reply by kangzoel** [reply]
Is that what you mean by subdirectory website?

**Reply by jrbecart** [reply]
Basically there are 3 types of websites:

- Root Domain, for example: laravel.com

- Subdomains, for example: : laravel.yourdomain.com

- Subdirectories or Subfolders , for example: :  yourdomain.com/laravel

> If you place the laravel folder under /path/to/web/directory/laravel (which is not recommended), it can be accessed in http://example.com/laravel/public

No, I access it straight with http://example.com/laravel 
Because for example with Nginx you set your root folder to the public folder of Laravel.

Nginx example: https://serversforhackers.com/c/nginx-php-in-subdirectory
Comparison of Subdomains vs subfolders from a SEO perspective: https://www.impactbnd.com/blog/subdomains-vs-subfolders

Subfolders is used everywhere, as you may know in education, within the same institution they are thousands of users with websites (Professors/Teachers/students, schools, faculties) You may have seen somewhere website like http://domain.com/~username , this mean account users have a website in the subfolder named with their username.
The main advantage of subfolder websites is that they are very easy to set up on the same domain, which is why they are popular.

**Reply by kangzoel** [reply]
I still don't get your question. You said Laravel doesn't support subdirectory website, but you are saying that you can access Laravel's website in subdirectory. Maybe you should make your question clearer. So that people can understand the problem

**Reply by ali-alharthi** [reply]
According to the links you provided what is looking for is something like this (not tested or recommended):

```
...
    root /var/www/my-domain.com/
...
    location / {
        try_files $uri $uri/ /index.php$is_args$args;
    }


    location /laravel {
        alias /var/www/my-domain.com/laravel/public;
        try_files $uri $uri/ @laravel;

        location ~ \.php$ {
            include snippets/fastcgi-php.conf;
            fastcgi_param SCRIPT_FILENAME $request_filename;
            fastcgi_pass unix:/var/run/php/php7.2-fpm.sock;
        }
    }
...
    location @laravel{
        rewrite /nested/(.*)$ /nested/index.php?/$1 last;
    }
...
```
Laravel does not care or know how you setup your webserver. However, there are things that are called best practices and this is not one of them imo!

Big websites tend to use one CMS/framework (WP for example) for everything they do landing-page, blog, news ...etc that handles routing such as your case! Or use sub-domain for each service if they use different frameworks.

**Reply by jrbecart** [reply]
Thanks @ali-alharthi ,
I already know how to do it and I am not looking for a solution here!

Can you explain why this "thing" is not standards or best practices?
Can you also please read my other comments where I explain the reason of the purpose of sub directories.

**Reply by ali-alharthi** [reply]
@jrbecart Sorry meant best practices and not standards.
The thing is rewrite rules tend to come at a cost especially if you are not experienced.
On the other hand, if you are an experienced devop I don't see laravel being a problem and you can manipulate the web server to do what you need it to do!

**Reply by jrbecart** [reply]
I am fully agree with you and the doc should be clearer on that point. Thanks!

**Reply by ramik5** [reply]
@jrbecart Hello! Can you tell me how to configure nginx so that the Laravel application works in a subdirectory?

I am using Ubuntu 20.04 + Laravel 8 in a Docker container.

My Nginx configuration:

```
server {

    charset utf-8;

    client_max_body_size 128M;

    server_name mydomain;

    index index.html;
    
    root /var/www/mydomain/www;
    
    error_log /var/www/mydomain/log/error.log;
 
    location / {
        proxy_pass http://127.0.0.1:9223;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_ssl_server_name on;
    }

    location ~ /\.(ht|svn|git) {
        deny all;
    }

    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/mydomain/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/mydomain/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}

server {
    if ($host = mydomain) {
        return 301 https://$host$request_uri;
    } # managed by Certbot



    server_name mydomain;

    listen 80;
    return 404; # managed by Certbot
}
```

The application works fine when I request mydomain/. Now, I want the application to be available at mydomain/subfolder. Thanks!

**Reply by jrbecart** [reply]
@ramik5 
For all my subsites I have some separate conf files.

In your first `server {}` declaration you can add (for example): `include /etc/nginx/config/*.conf;`
This load all sites configurations you have from the folder /etc/nginx/config/

Then, as an example, in the file /etc/nginx/config/example-other-site.conf  you can have something like this:

```
location ^~ /template {

    alias /var/www/example-template/public;
    try_files $uri $uri/ @template;
    
    error_log /var/log/nginx/template_error.log;
    access_log /var/log/nginx/template_access.log combined;

    location ~ \.php$ {
        fastcgi_pass php-upstream;
        include        fastcgi_params;
        fastcgi_param  SCRIPT_FILENAME $request_filename;
        fastcgi_param  SERVER_PORT 443; #important
        fastcgi_read_timeout 600;
    }
} 

location @template {
    rewrite /template/(.*)$ /template/index.php?/$1 last;
}
```

If you follow this configuration, in your case you will have
Site 1: https://mydomain (folder on your server: /var/www/mydomain/www)
Site 2: https://mydomain/template (folder on your server: /var/www/example-template/public)

In my config all my root folders end by /public, you may check that for your case.

Hope that helps!
